### DISCUSSION TITLE
Is it normal that you can download agents without beeing logged in?

### DISCUSSION THREAD

**Discussion by xardius** [Ylianst/MeshCentral#5551] [Q&A] [answered]
Hi.

On our installation of meshcentral (newest version on Debian) we have the problem that unknown clients are showing up.

While investigating we noted that the agents are available for everyone to download. The URL /meshagents is not reachable when you are not logged in but the direct link to the agent (e.g. /meshagents?id=3) works and starts the download of the agent without the need of being logged in. Is this normal behavior? Is it possible to disable this? We don't want and don't need this, we are using other mechanics to roll out our agents.

Thank you for your time :)

**Comment by si458** [comment]
https://ylianst.github.io/MeshCentral/meshcentral/faq/#help-ive-been-hacked-there-are-weird-agents-appearing-in-my-meshcentral-console

**Reply by dinger1986** [reply]
It most likely is that but also the statement is right, you can download agents without logging in and is there a way to disable that? I think its more a feature request to have the option to disable public links but dont know if that will break something else.

**Reply by NiceGuyIT** [reply]
The "Linux / BSD" script has the following. Downloading the agent does not need authentication. I agree with @dinger1986 this would be a good feature request to put the agent behind the token. Instead of this:
```bash
(
wget "https://mesh.example.com/meshagents?script=1" -O ./meshinstall.sh ||
wget "https://mesh.example.com/meshagents?script=1" --no-proxy -O ./meshinstall.sh
) &&
chmod 755 ./meshinstall.sh &&
sudo -E ./meshinstall.sh https://mesh.example.com '0000000000000000000000000000000000000000000000000000000000000000' ||
./meshinstall.sh https://mesh.example.com '0000000000000000000000000000000000000000000000000000000000000000'
```

Use something like this:
```bash
(
wget "https://mesh.example.com/meshagents?script=1&token=0000000000000000000000000000000000000000000000000000000000000000" -O ./meshinstall.sh ||
wget "https://mesh.example.com/meshagents?script=1&token=0000000000000000000000000000000000000000000000000000000000000000" --no-proxy -O ./meshinstall.sh
) &&
chmod 755 ./meshinstall.sh &&
sudo -E ./meshinstall.sh https://mesh.example.com '0000000000000000000000000000000000000000000000000000000000000000' ||
./meshinstall.sh https://mesh.example.com '0000000000000000000000000000000000000000000000000000000000000000'
```

I'll open a feature request.

**Reply by NiceGuyIT** [reply]
I don't know the level of effort to do this.
- #5554

**Reply by si458** [reply]
https://github.com/Ylianst/MeshCentral/issues/5554#issuecomment-1812571445

**Comment by NiceGuyIT** [comment, marked as answer]
@xardius You can block downloads by adding `"lockagentdownload": true` to the `config.json` and restarting the server. Be aware this will cause _scripted_ installs to fail; you will need to download the agent separately and modify the script accordingly.

```json
{
  "domains": {
    "": {
      "lockagentdownload": true
    }
  }
}
```

**Reply by si458** [reply]
i will still keep ur issue open about the token tho, as that would be an improvement/enhancement!

**Comment by wh1te909** [comment]
these are just generic agents, they don't have the server url or auth baked into them. so your unknown clients are not coming from this:

![mRemoteNG_Rt0jNSeihq](https://github.com/Ylianst/MeshCentral/assets/7434746/cd1e9819-4836-4478-a7a5-f0a197bbf37b)
